src/include/storage/index/hash_index.h

```diff
@@ -0,0 +1,269 @@
+
+    TERRIER_ASSERT(insert_result != uprase_result,
+                   "Either a new key was inserted (uprase_result), or the value already existed and a new value was "
+                   "inserted (insert_result).");
+
+    // Register an abort action with the txn context in case of rollback
+    txn->RegisterAbortAction(ERASE_KEY_ACTION);
+
+    return true;
+  }
+
+  bool InsertUnique(transaction::TransactionContext *const txn, const ProjectedRow &tuple,
+                    const TupleSlot location) final {
+    TERRIER_ASSERT(metadata_.GetSchema().Unique(), "This Insert is designed for indexes with uniqueness constraints.");
+    KeyType index_key;
+    index_key.SetFromProjectedRow(tuple, metadata_);
+    bool predicate_satisfied = false;
+
+    // The predicate checks if any matching keys have write-write conflicts or are still visible to the calling txn.
+    auto predicate = [txn](const TupleSlot slot) -> bool {
+      const auto *const data_table = slot.GetBlock()->data_table_;
+      const auto has_conflict = data_table->HasConflict(*txn, slot);
+      const auto is_visible = data_table->IsVisible(*txn, slot);
+      return has_conflict || is_visible;
+    };
+
+    bool UNUSED_ATTRIBUTE insert_result = false;
+
+    /**
+     * See the underlying container's API for more details, but the lambda below is invoked when the key is found.
+     *
+     * Captures:
+     * location the TupleSlot (value) being added to index
+     * insert_result captured by reference so it can be updated in outer scope. true if insert succeeded
+     * predicate_satisfied captured by reference so it can be updated in outer scope. true if uniqueness
+     * predicate was satisfied (i.e. there's a conflict with the insert)
+     * predicate std::function to evaluate for key uniqueness
+     *
+     * Args:
+     * value the current value for this key value (found by underlying containiner on lookup, then passed to
+     * key_found_fn)
+     *
+     * return true if cuckoohash_map's uprase_fn should delete the key/value pair. For inserts we always return false.
+     */
+    auto key_found_fn = [location, &insert_result, &predicate_satisfied, predicate](ValueType &value) -> bool {
+      if (std::holds_alternative<TupleSlot>(value)) {
+        const auto existing_location = std::get<TupleSlot>(value);
+        predicate_satisfied = predicate(existing_location);
+        if (!predicate_satisfied) {
+          // replace the TupleSlot with a ValueMap containing both the old and new inserted value
+          value = ValueMap({{location}, {existing_location}}, 2);
+          insert_result = true;
+        }
+      } else {
+        auto &value_map = std::get<ValueMap>(value);
+
+        // Check all of the existing TupleSlots at this key for conflicts
+        for (const auto i : value_map) {
+          predicate_satisfied = predicate_satisfied || predicate(i);
```

Comment:
Costly? Consider `std::any`/`std::all`?

Reply:
They're all O(n), but if it helps readability I can try to make it use the STL's stuff from <algorithm>

Reply:
Yep, but algorithm utils should have short-circuiting stuff by default.